### PR TITLE
feat [close #310]: Mounts system mountpoints like /home and /etc before systemd starts

### LINKED ISSUE #310: Systemd services get ignored
Since etc is mounted by systemd, user configuration of systemd (in /etc/systemd) is ignored. 

Stuff that I tried:
- mount luks drive without systemd (didn't work, probably because of udev, could maybe work somehow) (**this is the working solution, you can check out the PR**)
- mount luks drive with grub (I think grub only supports mounting the root drive, also, we would need to change the encryption method)
- running sysctrl daemon-reload and daemon-reexec during boot (it's just ignored for some reason. I guess this can't be used to regenerate the booting config)
- using soft reboot (kind of works but very unstable since systemd usually tries to unmount /etc while shutting down. Might work somehow) 

Dirty workaround solution:
If nothing else works, could also just copy /etc or at least /etc/systemd to the other root during an Operation (like an upgrade)
I don't like this since it's not intuitive for the user.

### PR DESCRIPTION
**Pull request by taukakao**
Fixes #310 

This command replaces the systemd mount units with a command that has to be run before systemd. 
This ensures that systemd recognizes the correct state of /etc.
This also makes the system more predictable, since there's not two different /etc folders to worry about.

To migrate to this version, the init system has to call the two new abroot functions before executing systemd.
The systemd mount units of old abroot versions don't cause any problems, since systemd doesn't remount them, so no futher migration is needed.

### CODE REVIEW

**Comment by taukakao** [comment]
Also, I can squash the commits if it's necessary but I think these commits are already pretty atomic.

**Review by kbdharun — APPROVED** [review]

File: cmd/unlock-var.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+			foundPart = true
+			break
+		}
+	}
+	if !foundPart {
+		return &VarInvalidError{varDisk}
+	}
+
+	uuid := varLuksPart.Uuid
+	cmdr.FgDefault.Println("/usr/sbin/cryptsetup luksOpen " + varDisk + " luks-" + uuid)
```
Comment (matbme) [inline cmd/unlock-var.go:129]: Do we really need to print the command?

File: cmd/unlock-var.go [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+	} else {
+		cryptsetupCmd := exec.Command("/usr/sbin/cryptsetup", "luksOpen", varDisk, "luks-"+uuid)
+		cryptsetupCmd.Stdin = os.Stdin
+		cryptsetupCmd.Stderr = os.Stderr
+		cryptsetupCmd.Stdout = os.Stdout
+		err := cryptsetupCmd.Run()
+		if err != nil {
+			return err
+		}
+		cmdr.Info.Println("The system mounts have been performed successfully.")
```
Comment (matbme) [inline cmd/unlock-var.go:142]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+	}
+	if !foundPart {
+		return &VarInvalidError{varDisk}
+	}
+
+	uuid := varLuksPart.Uuid
+	cmdr.FgDefault.Println("/usr/sbin/cryptsetup luksOpen " + varDisk + " luks-" + uuid)
+
+	if dryRun {
+		cmdr.Info.Println("Dry run complete.")
```
Comment (matbme) [inline cmd/unlock-var.go:132]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+	if err != nil {
+		cmdr.Error.Println(err)
+		os.Exit(1)
+	}
+	return nil
+}
+
+func unlockVar(cmd *cobra.Command, _ []string) error {
+	if !core.RootCheck(false) {
+		cmdr.Error.Println("You must be root to run this command.")
```
Comment (matbme) [inline cmd/unlock-var.go:85]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+			"perform a dry run of the operation",
+			false,
+		),
+	)
+
+	cmd.WithStringFlag(
+		cmdr.NewStringFlag(
+			"var-disk",
+			"m",
+			"pass /var disk directly instead of reading from configuration",
```
Comment (matbme) [inline cmd/unlock-var.go:61]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+		"",
+		"",
+		unlockVarCmd,
+	)
+
+	cmd.WithBoolFlag(
+		cmdr.NewBoolFlag(
+			"dry-run",
+			"d",
+			"perform a dry run of the operation",
```
Comment (matbme) [inline cmd/unlock-var.go:52]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+func (e *VarConfigError) Error() string {
+	return "reading the var disk from config is not implemented yet"
+}
+
+type VarInvalidError struct {
+	passedDisk string
+}
+
+func (e *VarInvalidError) Error() string {
+	return "the /var disk " + e.passedDisk + " does not exist"
```
Comment (matbme) [inline cmd/unlock-var.go:37]: This should be a translatable string.

File: cmd/unlock-var.go [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+	"github.com/spf13/cobra"
+
+	"github.com/vanilla-os/abroot/core"
+	"github.com/vanilla-os/orchid/cmdr"
+)
+
+type VarConfigError struct{}
+
+func (e *VarConfigError) Error() string {
+	return "reading the var disk from config is not implemented yet"
```
Comment (matbme) [inline cmd/unlock-var.go:29]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+		if err != nil {
+			return err
+		}
+		err = core.AtomicSwap(fstabFile, newFstabFile)
+		if err != nil {
+			return err
+		}
+		err = os.Rename(newFstabFile, fstabFile+".old")
+		if err != nil {
+			cmdr.Warning.Println("Old Fstab file will keep .new suffix")
```
Comment (matbme) [inline cmd/mount-sys.go:268]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+	cmdr.FgDefault.Println("Adding line: ", currentRootLine)
+
+	linesNew = append([]string{currentRootLine}, linesNew...)
+
+	newFstabContents := strings.Join(linesNew, "\n")
+
+	newFstabFile := fstabFile + ".new"
+
+	if !dryRun {
+		cmdr.FgDefault.Println("writing new fstab file")
```
Comment (matbme) [inline cmd/mount-sys.go:257]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+			linesNew = append(linesNew, line)
+			continue
+		}
+
+		cmdr.FgDefault.Println("Deleting line: ", line)
+	}
+
+	currentRootLine := "UUID=" + uuid + " / btrfs defaults 0 0"
+
+	cmdr.FgDefault.Println("Adding line: ", currentRootLine)
```
Comment (matbme) [inline cmd/mount-sys.go:248]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+			linesNew = append(linesNew, line)
+			continue
+		}
+
+		if !slices.Contains[[]string](systemMounts, words[1]) {
+			linesNew = append(linesNew, line)
+			continue
+		}
+
+		cmdr.FgDefault.Println("Deleting line: ", line)
```
Comment (matbme) [inline cmd/mount-sys.go:243]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 13]
```diff
@@ -0,0 +1,274 @@
...
+	"github.com/vanilla-os/abroot/settings"
+	"github.com/vanilla-os/orchid/cmdr"
+)
+
+type DiskLayoutError struct {
+	Device string
+}
+
+func (e *DiskLayoutError) Error() string {
+	return fmt.Sprintf("device %s has an unsupported layout", e.Device)
```
Comment (matbme) [inline cmd/mount-sys.go:35]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 14]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+type PartNotFoundError struct {
+	Partition string
+}
+
+func (e *PartNotFoundError) Error() string {
+	return fmt.Sprintf("partition %s could not be found", e.Partition)
```
Comment (matbme) [inline cmd/mount-sys.go:43]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+		"",
+		"",
+		mountSysCmd,
+	)
+
+	cmd.WithBoolFlag(
+		cmdr.NewBoolFlag(
+			"dry-run",
+			"d",
+			"perform a dry run of the operation",
```
Comment (matbme) [inline cmd/mount-sys.go:58]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+	if err != nil {
+		cmdr.Error.Println(err)
+		os.Exit(1)
+	}
+	return nil
+}
+
+func mountSys(cmd *cobra.Command, _ []string) error {
+	if !core.RootCheck(false) {
+		cmdr.Error.Println("This operation requires root.")
```
Comment (matbme) [inline cmd/mount-sys.go:82]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+		return &PartNotFoundError{"current root"}
+	}
+	err = adjustFstab(present.Uuid, dryRun)
+	if err != nil {
+		cmdr.Error.Println(err)
+		os.Exit(8)
+	}
+
+	if dryRun {
+		cmdr.Info.Println("Dry run complete.")
```
Comment (matbme) [inline cmd/mount-sys.go:128]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 18]
```diff
@@ -0,0 +1,274 @@
...
+	} else {
+		cmdr.Info.Println("The system mounts have been performed successfully.")
```
Comment (matbme) [inline cmd/mount-sys.go:130]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	return nil
+}
+
+func mountVar(varPart core.Partition, dryRun bool) error {
+	cmdr.FgDefault.Println("mounting " + varPart.Device + " in /var")
```
Comment (matbme) [inline cmd/mount-sys.go:137]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+
+	overlays := []overlayMount{
+		{"/.system/etc", []string{"/.system/etc"}, "/var/lib/abroot/etc/" + rootLabel, "/var/lib/abroot/etc/" + rootLabel + "-work"},
+	}
+
+	for _, overlay := range overlays {
+		lowerCombined := strings.Join(overlay.lowerdirs, ":")
+		options := "lowerdir=" + lowerCombined + ",upperdir=" + overlay.upperdir + ",workdir=" + overlay.workdir
+
+		cmdr.FgDefault.Println("mounting overlay mount " + overlay.destination + " with options " + options)
```
Comment (matbme) [inline cmd/mount-sys.go:193]: This should be a translatable string.

File: cmd/mount-sys.go [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func adjustFstab(uuid string, dryRun bool) error {
+	cmdr.FgDefault.Println("switching the root in fstab")
```
Comment (matbme) [inline cmd/mount-sys.go:207]: This should be a translatable string.

**Review by matbme — CHANGES_REQUESTED** [review]
Implementation looks good. My spam of suggestions is mostly related to missing translations.

**Review by mirkobrombin — APPROVED** [review]
LGTM. I agree with Mat about the missing translations, but per our conversation on discord, should be fine due to the command being strictly internal.

Anyway, adding the translation keys should not be so hard, so I'll leave the decision to you.

File: cmd/unlock-var.go (hunk 1 above)
Comment (taukakao) [reply cmd/unlock-var.go:129]: Oh yeah, we don't. 
It's a relic from my testing.

**Comment by taukakao** [comment]
As discussed on Discord, I don't want to put translations in.
These commands are not user-facing so they don't need to be translated. It would only put extra work on the translators without making a difference.

**Comment by mirkobrombin** [comment]
Not to go over Mat but given the discussions on Discord regarding the translations and given the resolution of the discussions, I still proceed with the merge without his confirmation.

**Pull request merged**